src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -365,10 +365,15 @@ static LanguageServer.Protocol.Diagnostic ConvertToLspDiagnostic(DiagnosticData
 
         private bool IncludeDiagnostic(DiagnosticData diagnostic)
         {
+            // Some diagnostics only apply to certain clients, e.g. Razor.
             if (!diagnostic.Properties.TryGetValue(nameof(DocumentPropertiesService.DiagnosticsLspClientName), out var diagnosticClientName))
             {
-                // This diagnostic is not restricted to a specific LSP client, so just pass it through
-                return true;
+                // At this point, we know that the diagnostic we're working with is contained in a C#/VB file, since
+                // if we were working with a non-C#/VB file, then DocumentPropertiesService.DiagnosticsLspClientName
+                // should have been populated.
+                // Thus, if we have anything other than a null client name, we don't want to publish the diagnostic
+                // (since a null client name represents the C#/VB language client).
+                return _clientName == null;
```

Comment:
There might be a GetOrDefault extension for dictionary which returns null if key not found, then we could just do '''return diagnostic.Properties.GetOrDefault() == _clientName'''' since basically the same check is being done in and outside of the if.